presto-postgresql/src/main/java/com/facebook/presto/plugin/postgresql/PostgreSqlClient.java

```diff
@@ -87,7 +95,29 @@ protected String toSqlType(Type type)
         if (VARBINARY.equals(type)) {
             return "bytea";
         }
+        if (JSON.equals(type)) {
+            return "jsonb";
+        }
 
         return super.toSqlType(type);
     }
+
+    @Override
+    public Optional<ReadMapping> toPrestoType(ConnectorSession session, JdbcTypeHandle typeHandle)
+    {
+        if (typeHandle.getJdbcType() == Types.OTHER) {
+            switch (typeHandle.getJdbcTypeName()) {
+                case "jsonb":
+                case "json":
+                    return Optional.of(jsonReadMapping());
+            }
+        }
+
+        return super.toPrestoType(session, typeHandle);
+    }
+
+    private static ReadMapping jsonReadMapping()
+    {
+        return sliceReadMapping(JSON, (resultSet, columnIndex) -> utf8Slice(resultSet.getString(columnIndex)));
```

Comment:
@electrum with JsonType not being in SPI, how can a connector know this is the correct runtime representation of the JsonType?
We certainly don't need to move JsonType to SPI, but we probably need to have some SPI factory method (or other SPI-level contract) for this.

Reply:
My first thought was to resolve the CAST operator using TypeManager. However, cast from varchar to json is actually different from the type constructor. Maybe we should add a method to lookup the type constructor?

@martint thoughts?

Reply:
Sill question -- why is the cast from varchar different from type constructor? Is't not intuitive.

Reply:
Casting a `varchar` to JSON results in a JSON string. The type constructor takes a JSON literal and is equivalent to `json_parse()`. I agree it's confusing, but my memory is that it works like that so that it's consistent when done recursively. For example, casting `varchar` and `array(varchar)` work the same way.

See here for a bit more detail: https://prestodb.io/docs/current/functions/json.html#json_parse

Reply:
I just put up a PR to allow resolving the type constructor: https://github.com/prestodb/presto/pull/12007

Reply:
@electrum Can you explain how can I use it from PostgreSqlClient ?
Currently I am using a copied version of jsonParse  (```jsonParse(utf8Slice(resultSet.getString(columnIndex))))```) since it's not exposed in the spi. Same for JsonType that is needed by ```com.facebook.presto.plugin.jdbc.ReadMapping```